Add POST /content/{siteId}/folder API v2

phuongnq opened the pull request:
### Ticket reference or full description of what's in the PR
https://github.com/craftercms/craftercms/issues/6061

From
```
POST /api/1/services/api/1/content/create-folder.json
```

To
```
POST /api/2/content/{siteId}/folder
```

phuongnq committed:
Add POST /content/{site}/folder API v2

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/service/content/ContentServiceImpl.java:
```diff
@@ -331,6 +336,17 @@ public Resource getContentAsResource(@ValidateStringParam String site,
+    @Override
+    @Valid
+    @ValidateAction(type = Type.CREATE)
+    @HasPermission(type = DefaultPermission.class, action = PERMISSION_CONTENT_WRITE)
```
What about `folder_create` permission?

jmendeza replied:
@phuongnq let's use `folder_create`.
Let's also rename it to be `create_folder` so it is consistent with the permission names we have been using

phuongnq replied:
Updated to use `create_folder`. Thank you!
Note: We may need to review the PR when merging this to make sure the pipeline is up-to-date with other PRs which modify the same.

phuongnq committed:
Update folder_create permission to create_folder

jmendeza commented on src/main/api/2/studio-api-2.yaml:
```diff
@@ -3682,6 +3682,55 @@ paths:
         '500':
           $ref: '#/components/responses/InternalServerError'
 
+  /content/create_folder:
```
@phuongnq please refer to the API v2 document. This would now be 
POST /content/{siteId}/folder

phuongnq replied:
I updated to new API endpoint. Thank you!

phuongnq committed:
Update response status in case of creating folder failed

jmendeza commented on src/main/java/org/craftercms/studio/controller/rest/v2/ContentController.java:
```diff
@@ -297,7 +298,15 @@ public ResultList<ItemVersion> getHistory(@ValidSiteId @RequestParam(value = REQ
+                                  @Valid @RequestBody CreateFolderRequestBody body) throws UserNotFoundException, ServiceLayerException {
+        boolean succeeded = contentService.createFolder(siteId, body.getPath(), body.getName());
+        var result = new ResultOne();
+        result.setResponse(succeeded ? OK : FOLDER_CREATION_FAILED);
```
Let's return 201 on success

phuongnq replied:
We have the CREATED APIResponse, I updated to respond with this value.
```
Status Code 200
{
    "response": {
        "code": 1,
        "message": "Created",
        "remedialAction": "",
        "documentationUrl": ""
    }
}
```

jmendeza commented on src/main/java/org/craftercms/studio/api/v2/service/content/ContentService.java:
```diff
@@ -209,4 +209,15 @@ boolean renameContent( String site, String path, String name)
+     * @throws UserNotFoundException user not found error
+     * @throws ServiceLayerException general service error
+     */
+    boolean createFolder(String siteId, String path, String name) throws ServiceLayerException, UserNotFoundException;
```
Let's make this void and throw an exception on failure.

phuongnq replied:
I updated to throw exceptions on failure

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/service/content/ContentServiceImpl.java:
```diff
@@ -365,6 +369,16 @@ public List<ItemVersion> getContentVersionHistory(@SiteId String siteId, @Protec
         return contentServiceInternal.getContentVersionHistory(siteId, path);
     }
 
+    @Valid
```
Let's also validate site state
`@RequireSiteReady`

phuongnq replied:
Added `@RequireSiteReady`

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/service/content/ContentServiceImpl.java:
```diff
@@ -365,6 +369,16 @@ public List<ItemVersion> getContentVersionHistory(@SiteId String siteId, @Protec
+    @Valid
+    @ValidateAction(type = Type.CREATE)
+    @HasPermission(type = DefaultPermission.class, action = PERMISSION_CREATE_FOLDER)
+    public boolean createFolder(@ProtectedResourceId(SITE_ID_RESOURCE_ID) @ValidateStringParam @SiteId String siteId,
```
- `@ValidateStringParam` without parameters does nothing. Consider removing it
- `@SiteId` now implies `@ProtectedResourceId(SITE_ID_RESOURCE_ID)`
- Syntax/format validations should happen in controller method when validating `CreateFolderRequestBody`

phuongnq replied:
Updated as of above.

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/service/content/internal/ContentServiceInternalImpl.java:
```diff
@@ -292,10 +300,80 @@ public List<ItemVersion> getContentVersionHistory(String siteId, String path) th
+        String folderPath = path + FILE_SEPARATOR + name;
+        boolean toRet = false;
+        String commitId = contentRepository.createFolder(siteId, path, name);
+        if (commitId != null) {
```
ContentRepository.createFolder should never return null. Instead, it should throw an Exception if an error happens

phuongnq replied:
I updated to throw exceptions instead of returning `null`

jmendeza commented on src/main/java/org/craftercms/studio/api/v2/dal/SiteDAO.java:
```diff
@@ -66,4 +65,12 @@ public interface SiteDAO {
-}
+
+    /**
+     * update last commit id
```
Update (with initial capital)

phuongnq replied:
Updated

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/service/content/internal/ContentServiceInternalImpl.java:
```diff
@@ -292,10 +300,80 @@ public List<ItemVersion> getContentVersionHistory(String siteId, String path) th
         }
     }
 
```
`@Override` missing?

phuongnq replied:
Updated

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/repository/GitContentRepository.java:
```diff
@@ -1354,6 +1354,57 @@ public boolean contentExists(String site, String path) {
 
+    @Override
+    public String createFolder(String site, String path, String name) {
+        String commitId = null;
```
Please refactor this method to never return null. Throw an exception if there is any error

phuongnq replied:
I updated to throw exceptions instead of returning null

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/service/content/internal/ContentServiceInternalImpl.java:
```diff
@@ -27,12 +27,14 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.craftercms.studio.api.v1.service.configuration.ServicesConfig;
 import org.craftercms.studio.api.v1.service.security.SecurityService;
```
Check if this can be replaced with v2 SecurityService.
In general, let's try to replace V1 service calls as much as possible

phuongnq replied:
I added a new class `org.craftercms.studio.impl.v2.utils.ContentUtils`

jmendeza replied:
@phuongnq can we replace this `import org.craftercms.studio.api.v1.service.security.SecurityService;` with the v2 version of it?

phuongnq committed:
throw exception if create folder failed

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/repository/GitContentRepository.java:
```diff
@@ -1354,6 +1354,57 @@ public boolean contentExists(String site, String path) {
+            }
+
+            return commitId;
+        } catch (Exception e) {
```
Please add a catch for `ServiceLayerException` so we just rethrow the ones from the try block

phuongnq replied:
I added a catch for `ServiceLayerException`

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/repository/GitContentRepository.java:
```diff
@@ -1354,6 +1354,57 @@ public boolean contentExists(String site, String path) {
+            return commitId;
+        } catch (Exception e) {
+            logger.error("Failed to add file to git in site '{}' path '{}'", siteId, newFolderPath, e);
+            throw new ServiceLayerException(format("Error creating new folder at path '{}' site '{}'", newFolderPath, siteId));
```
Include the caught exception `e` in the new ServiceLayerException constructor

phuongnq replied:
Updated

jmendeza commented on src/main/java/org/craftercms/studio/impl/v2/repository/blob/BlobAwareContentRepository.java:
```diff
@@ -263,21 +265,22 @@ public String writeContent(String site, String path, InputStream content) throws
-            return null;
+            logger.error("Failed to create folder in site '{}' path '{}'", siteId, path, e);
+            Path newFolderPath = Paths.get(path, name);
+            throw new ServiceLayerException(format("Error creating new folder at path '{}' site '{}'", newFolderPath, siteId));
```
Include the caught Exception `e` in the new ServiceLayerException constructor

phuongnq replied:
Updated

jmendeza commented on src/main/api/2/studio-api-2.yaml:
```diff
@@ -3792,6 +3792,60 @@ paths:
+                - path
+                - name
+      responses:
+        '200':
```
This should be 201 CREATED

phuongnq replied:
Updated to response with 201 status

phuongnq committed:
Use SecurityService v2 in ContentServiceInternalImpl

phuongnq commented:
@jmendeza I refactored to use `org.craftercms.studio.api.v2.service.security.SecurityService`. There are some circular reference so I also updated them for SecurityService, Group Internal, User Internal

jmendeza commented on src/main/java/org/craftercms/studio/api/v2/service/security/SecurityService.java:
```diff
@@ -56,4 +56,12 @@ long getAvailableActions(String username, String site, String path)
+     * @return list of groups
+     * @throws ServiceLayerException
+     */
+    List<String> getSiteGroups(String siteId) throws ServiceLayerException;
```
@phuongnq Why are we moving this from GroupService? Is it because of the circular dependencies you mentioned?

phuongnq replied:
@jmendeza Right, we have several beans dependent on each other which caused the circular dependencies on the User and Group Impl. Even the name is `getSiteGroups`, this method just reads the XML configuration and returns the list which we do similar things in `SecurityServcice` for permission so I think this is a good candidate to move the method into.

jmendeza reviewed (changes requested):
@phuongnq please update the base branch so these changes go to `develop`. (it is pointing to `dev/4.2.x`)

phuongnq commented:
Close in favor of https://github.com/craftercms/studio/pull/3382
